### ISSUE TITLE
JWTResponseListener / Request is null

### ISSUE THREAD

**Issue by Wonsy** [lexik/LexikJWTAuthenticationBundle#244]
Hello everyone,

I just updated bundle, and my listener which was previously working doesn't anymore.

I ran onto issue #128, which gave me the info that Request system has been changed in 1.4.

Here's a snipped of my listener : 

```
public function onAuthenticationSuccessResponse(AuthenticationSuccessEvent $event)
{
       $data    = $event->getData();
        $user    = $event->getUser();
       $request = $event->getRequest();
```

$request used to be filled, suddenly gone to NULL after update.

So, tracked down to this point : 

Security\Http\Authentication\AuthenticationSuccessHandler.php : 44

```
 public function onAuthenticationSuccess(Request $request, TokenInterface $token)
{
         $user = $token->getUser();
         $jwt  = $this->jwtManager->create($user);
 
         $response = new JsonResponse();
         $event    = new AuthenticationSuccessEvent(['token' => $jwt], $user, null, $response);
```

Changed to : 

```
public function onAuthenticationSuccess(Request $request, TokenInterface $token)
{
         $user = $token->getUser();
         $jwt  = $this->jwtManager->create($user);
 
         $response = new JsonResponse();
         $event    = new AuthenticationSuccessEvent(['token' => $jwt], $user, $request, $response);

```

solved the problem. However, I feel (very) uncomfortable about this fix : as stated everywhere in your code, this is deprecated and soon removed; and I really don't want a custom version of a bundle (eek.)

So, I was wondering if : 
A°) "It's not a bug, it's a feature" ? (or is it, really ?)
B°) If it's a feature, I didn't see anywhere in docs a way to get back my request data in my response listener ?

I didn't found either any closed issue on this topic, am I the only one ? (really ?)

Juste one last thing : it's my first issue opened on Github... please excuse me if something is wrong with it (any advice is appreciated !)

Anyway : thanks for your bundle !

( Edit : had some troubles with code snipped markup... switched to "quote" one. )

**Comment by chalasr** [comment]
@Wonsy First, thank you very much for opening this issue, even more if it is your first opened issue on github.

Thanks to this issue, I realized we introduced a BC break in #200. 
It should be fixed now. Could you please try to update the bundle (the fix being tagged as `v1.7.1') and confirm that all work as before from your side?

> So, I was wondering if : 
A°) "It's not a bug, it's a feature" ? (or is it, really ?)
B°) If it's a feature, I didn't see anywhere in docs a way to get back my request data in my response listener 

Well, it's indeed a feature at first, which removed a bad practice from our code:
You should consider to avoid retrieving the `Request` object from the event itself.

Explanation:
Since symfony 2.4, it's deprecated to inject the `@request` service and a bad practice to inject some `Request` instances in your classes.
The reason is that a `RequestStack` class and its `@request_stack` service have been introduced in the same time, representing the right way to access the current request from anywhere in your code.

In the `1.x` releases  of this bundle, pretty much all `Event` classes take a `Request` instance as constructor argument, allowing users to use this request in their listeners, as you are doing.
In `2.0 we grabbed the opportunity to avoid this practice by removing those `Request` injections from our `Event` and `EventListener` classes (and some other).

Here is what you need to do for avoiding this practice (and btw have a smooth upgrade to the `2.0` version of this bundle which should be released this week, with a bunch of nice new features).

 __Before__
  
  ```yaml
  services:
     app.jwt_authentication_success_listener:
         class: AppBundle\EventListener\JWTAuthenticationSuccessListener
         tags:
             - { name: kernel.event_listener, event: lexik_jwt_authentication.on_authentication_success, method: onAuthenticationSuccessResponse }
  ```
  
  ```php
  use Lexik\Bundle\JWTAuthenticationBundle\Event\AuthenticationSuccessEvent;
  
  class JWTAuthenticationSuccessListener
  {
      public function onAuthenticationSuccessResponse(AuthenticationSuccessEvent $event)
      {
          $request = $event->getRequest();
          // ...
      }
  }
  ```
  
  __After__
  
  ```yaml
  services:
      app.jwt_authentication_success_listener:
          class: AppBundle\EventListener\JWTAuthenticationSuccessListener
          arguments: [ '@request_stack' ]
          tags:
              - { name: kernel.event_listener, event: lexik_jwt_authentication.on_jwt_created, method: onJWTCreated }
  ```
  
  ```php  
  use Lexik\Bundle\JWTAuthenticationBundle\Event\AuthenticationSuccessEvent;
  use Symfony\Component\HttpFoundation\RequestStack;

  class JWTAuthenticationSuccessListener
  {
      private $requestStack;
      
      public function __construct(RequestStack $requestStack)
      {
          $this->requestStack = $requestStack;
      }
      
      public function onAuthenticationSuccessResponse(AuthenticationSuccessEvent $event)
      {
          $request = $this->requestStack->getCurrentRequest();
      }
  }
  ```

This way you can get the same `Request` instance from many event listeners, just need to inject the `@request_stack` service and call `RequestStack::getCurrentRequest()`.

You can see all the upgrade path for `2.0` in [the `UPGRADE.md` file of the `2.0` branch of this bundle](https://github.com/lexik/LexikJWTAuthenticationBundle/blob/2.0/UPGRADE.md), which will become `master` when released. 

I hope you can easily make the change in your code, in order to have a smooth upgrade to 2.0 as soon as this one will be available.
Please do not hesitate to give feedbacks, it will be really appreciated. And if this solve your problem, please close the issue :).

Again, thank you.

**Comment by chalasr** [comment]
Closing as fixed
